### PR TITLE
Add ci.pipeline.run.duration metric

### PR DESCRIPTION
**Pull request by cyrille-leclerc**
* Add `ci.pipeline.run.duration{ci.pipeline.id=xxx, ci.pipeline.run.result=yyy}` histogram metric
   * Config params `otel.instrumentation.jenkins.run.metric.duration.allow_list` and `otel.instrumentation.jenkins.run.metric.duration.allow_list`: Java regex strings
* Cleanup some docs

### Submitter checklist
- [ ] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [ ] Ensure that the pull request title represents the desired changelog entry
- [ ] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java [hunk 1]
```diff
@@ -104,42 +115,60 @@ public void postConstruct() {
 
         // METRICS
         activeRunGauge = new AtomicInteger();
+
+        runDurationHistogram = meter.histogramBuilder("ci.pipeline.run.duration")
```
Comment (christophe-kamphaus-jemmic) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:119]: How are the buckets for the histogram defined?

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java [hunk 2]
```diff
@@ -395,6 +423,19 @@ public void _onFinalized(@NonNull Run run) {
...
+                runDurationHistogramAllowList.matcher(run.getParent().getFullName()).matches()
+                    &&
+                    !runDurationHistogramDenyList.matcher(run.getParent().getFullName()).matches() ?
+                    run.getParent().getFullName() : "#other#";
+            runDurationHistogram.record(
+                TimeUnit.SECONDS.convert(run.getDuration(), TimeUnit.MILLISECONDS),
+                Attributes.of(
+                    JenkinsOtelSemanticAttributes.CI_PIPELINE_ID, pipelineId,
+                    JenkinsOtelSemanticAttributes.CI_PIPELINE_RUN_RESULT, result.toString(),
+                    JenkinsOtelSemanticAttributes.CI_PIPELINE_RUN_COMPLETED, result.isCompleteBuild())
```
Comment (christophe-kamphaus-jemmic) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:437]: `result.isCompleteBuild()==false` means that the run was aborted.
It makes sense to count the aborted jobs separately. :+1:

However aren't aborted jobs already covered by the `result==Result.ABORTED` (`CI_PIPELINE_RUN_RESULT` attribute)?
Do we need a separate completed attribute?

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java (hunk 1 above)
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:119]:
```
        runDurationHistogram = meter.histogramBuilder("ci.pipeline.run.duration")
        // TODO clarify histogram buckets.
```
Great catch Christophe, I forgot to add a TODO to clarify buckets

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java (hunk 2 above)
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:438]:
```suggestion
                    // TODO do we need this `completed` dimension that we captured on spans? Can't this inferred from the `result` attribute ?
                    JenkinsOtelSemanticAttributes.CI_PIPELINE_RUN_COMPLETED, result.isCompleteBuild())
```

> However aren't aborted jobs already covered by the result==Result.ABORTED (CI_PIPELINE_RUN_RESULT attribute)?

It was recopied from what we put in the span. Question added a a todo

**Comment by cyrille-leclerc** [comment]
@miraccan00 could you review this PR please?

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java [hunk 3]
```diff
@@ -104,42 +125,75 @@ public void postConstruct() {
...
 
+    @Override
+    public void afterConfiguration(ConfigProperties configProperties) {
+        Pattern newRunDurationHistogramAllowList;
+        Pattern newRunDurationHistogramDenyList;
+        try {
+            newRunDurationHistogramAllowList = Optional
+                .ofNullable(configProperties.getString(JenkinsOtelSemanticAttributes.OTEL_INSTRUMENTATION_JENKINS_RUN_DURATION_ALLOW_LIST))
+                .map(Pattern::compile)
+                .orElse(MATCH_NOTHING);
```
Comment (christophe-kamphaus-jemmic) [inline src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:181]: Does this mean that by default the `ci.pipeline.run.duration` metric will count all job runs under `#other#` until an admin has configured the `otel.instrumentation.jenkins.run.metric.duration.allow_list` property?

File: docs/monitoring-metrics.md [hunk 4]
```diff
@@ -24,7 +24,21 @@ or APIs ([here](https://www.elastic.co/guide/en/kibana/current/dashboard-import-
...
+* Type: Histogram with buckets: `1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192` (buckets subject to change)
+* Unit: `s`
+* Attributes:
+  * `ci.pipeline.id`: The full name of the Jenkins job if complying with the allow and deny lists specified through 
+    configuration parameters documented below, otherwise `#other#` to limit the cardinality of the metric. 
+  Example: `my-team/my-app/main`. See `hudson.model.AbstractItem#getFullName()`.
+  * `ci.pipeline.result`: `SUCCESS`, `UNSTABLE`, `FAILUIRE`, `NOT_BUILT`, `ABORTED`. See `hudson.model.Run#getResult()`.
+* Configuration parameters to control the cardinality of the `ci.pipeline.id` attribute:
+  * `otel.instrumentation.jenkins.run.metric.duration.allow_list`: Java regex. Example `jenkins_folder_a/.*|jenkins_folder_b/.*`
+  * `otel.instrumentation.jenkins.run.metric.duration.deny_list`: Java regex. Example `.*test.*`
```
Comment (christophe-kamphaus-jemmic) [inline docs/monitoring-metrics.md:39]: Maybe specify the defaults for these properties?

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java (hunk 3 above)
Comment (christophe-kamphaus-jemmic) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:181]: It's a very good idea. That way all builds are taken into account for the metric while still keeping cardinality under control.
It also allows writing queries to get statistics for all builds on a Jenkins instance (by aggregating away the `ci.pipeline.id`).

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java [hunk 5]
```diff
@@ -72,17 +81,29 @@
 @Extension(dynamicLoadable = YesNoMaybe.YES, optional = true)
 public class MonitoringRunListener extends OtelContextAwareAbstractRunListener implements OpenTelemetryLifecycleListener {
 
+    static final Pattern MATCH_ANYTHING = Pattern.compile(".*");
+    static final Pattern MATCH_NOTHING = Pattern.compile("$^");
+
+    static final List<Double> DURATION_SECONDS_BUCKETS =
+        unmodifiableList(
+            asList(1D, 2D, 4D, 8D, 16D, 32D, 64D, 128D, 256D, 512D, 1024D, 2048D, 4096D, 8192D));
```
[the remark this answers is not in the document]
Comment (christophe-kamphaus-jemmic) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:89]: I do not think we need to start below `1s`.
I could imagine that users might want to customize the buckets (eg. if they want higher resolution around the usual duration of their pipelines).
There might also be users with jobs running extra long.

What do you think of making it configurable?
(these buckets look like good defaults)

File: docs/monitoring-metrics.md [hunk 6]
```diff
@@ -24,7 +24,26 @@ or APIs ([here](https://www.elastic.co/guide/en/kibana/current/dashboard-import-
 |------------------------------------------------|----------------------------------|
 | <img alt="Jenkins Health Dashboard with Elastic Kibana" width="300px" src="https://raw.githubusercontent.com/jenkinsci/opentelemetry-plugin/master/docs/images/kibana_jenkins_overview_dashboard.png" /> | <img alt="Jenkins Agent Provisioning Health Dashboard with Elastic Kibana" width="300px" src="https://raw.githubusercontent.com/jenkinsci/opentelemetry-plugin/master/docs/images/kibana_jenkins_provisioning_dashboard.png" /> |
 
-## Jenkins Health Metrics
+## Build Duration
+
+**⚠️ In order to control metrics cardinality, the `ci.pipeline.run.duration` metrics are enabled by default 
+aggregating the durations of all the jobs/pipelines under the umbrella `ci.pipeline.id=#other#`.
+To enable per  job/pipeline metrics, use the allow and deny list setting the configuration parameters 
+`otel.instrumentation.jenkins.run.metric.duration.allow_list` and `otel.instrumentation.jenkins.run.metric.duration.deny_list`.**
```
Comment (cyrille-leclerc) [inline docs/monitoring-metrics.md:32]: @christophe-kamphaus-jemmic please review

File: docs/monitoring-metrics.md [hunk 7]
```diff
@@ -24,7 +24,26 @@ or APIs ([here](https://www.elastic.co/guide/en/kibana/current/dashboard-import-
...
+* Type: Histogram with buckets: `1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192` (buckets subject to change)
+* Unit: `s`
+* Attributes:
+  * `ci.pipeline.id`: The full name of the Jenkins job if complying with the allow and deny lists specified through 
+    configuration parameters documented below, otherwise `#other#` to limit the cardinality of the metric. 
+  Example: `my-team/my-app/main`. See `hudson.model.AbstractItem#getFullName()`.
+  * `ci.pipeline.result`: `SUCCESS`, `UNSTABLE`, `FAILUIRE`, `NOT_BUILT`, `ABORTED`. See `hudson.model.Run#getResult()`.
+* Configuration parameters to control the cardinality of the `ci.pipeline.id` attribute:
+  * `otel.instrumentation.jenkins.run.metric.duration.allow_list`: Java regex, default value: `$^` (ie match nothing). Example `jenkins_folder_a/.*|jenkins_folder_b/.*`
+  * `otel.instrumentation.jenkins.run.metric.duration.deny_list`: Java regex, default value: `$^` (ie match nothing). Example `.*test.*`
```
Comment (cyrille-leclerc) [inline docs/monitoring-metrics.md:44]: @christophe-kamphaus-jemmic please review

File: docs/monitoring-metrics.md (hunk 4 above)
Comment (cyrille-leclerc) [reply docs/monitoring-metrics.md]: Please review updated docs. I tagged you on the improved sections that describe this behavior.

File: docs/monitoring-metrics.md [hunk 8]
```diff
@@ -24,7 +24,26 @@ or APIs ([here](https://www.elastic.co/guide/en/kibana/current/dashboard-import-
 |------------------------------------------------|----------------------------------|
 | <img alt="Jenkins Health Dashboard with Elastic Kibana" width="300px" src="https://raw.githubusercontent.com/jenkinsci/opentelemetry-plugin/master/docs/images/kibana_jenkins_overview_dashboard.png" /> | <img alt="Jenkins Agent Provisioning Health Dashboard with Elastic Kibana" width="300px" src="https://raw.githubusercontent.com/jenkinsci/opentelemetry-plugin/master/docs/images/kibana_jenkins_provisioning_dashboard.png" /> |
 
-## Jenkins Health Metrics
+## Build Duration
+
+**⚠️ In order to control metrics cardinality, the `ci.pipeline.run.duration` metrics are enabled by default 
+aggregating the durations of all the jobs/pipelines under the umbrella `ci.pipeline.id=#other#`.
+To enable per  job/pipeline metrics, use the allow and deny list setting the configuration parameters
```
Comment (christophe-kamphaus-jemmic) [inline docs/monitoring-metrics.md:31]:
```suggestion
To enable per job/pipeline metrics, use the allow and deny list setting the configuration parameters 
```

File: docs/monitoring-metrics.md (hunk 7 above)
Comment (christophe-kamphaus-jemmic) [reply docs/monitoring-metrics.md:44]: LGTM

File: docs/monitoring-metrics.md (hunk 6 above)
Comment (christophe-kamphaus-jemmic) [reply docs/monitoring-metrics.md:32]: LGTM (left a small suggested change comment)

File: docs/monitoring-metrics.md (hunk 4 above)
Comment (christophe-kamphaus-jemmic) [reply docs/monitoring-metrics.md]: The updated docs look good to me.

File: src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:91]: > What do you think of making it configurable?
@christophe-kamphaus-jemmic 💯 . Please review the TODO comment I added above

[the remark this answers is not in the document]
Comment (christophe-kamphaus-jemmic) [reply src/main/java/io/jenkins/plugins/opentelemetry/job/MonitoringRunListener.java:91]: LGTM

File: src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsSemanticMetrics.java [hunk 9]
```diff
@@ -7,6 +7,7 @@
 
 public class JenkinsSemanticMetrics {
     public static final String  CI_PIPELINE_RUN_ACTIVE =            "ci.pipeline.run.active";
+    public static final String  CI_PIPELINE_RUN_DURATION =          "ci.pipeline.run.duration";
```
Comment (christophe-kamphaus-jemmic) [inline src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsSemanticMetrics.java:10]: This matches the other metrics currently defined in this plugin (use of `ci.*` namespace).
I've checked against https://opentelemetry.io/docs/specs/semconv/attributes-registry/cicd/ and https://github.com/open-telemetry/semantic-conventions/blob/e63b2cfd377688c3f0a9054636fb52d973998a7a/docs/cicd/cicd-metrics.md from PR https://github.com/open-telemetry/semantic-conventions/pull/1383 to verify if that name is likely to be accepted in the OpenTelemetry semantic conventions.
Currently it looks more likely that the `cicd.*` namespace will be used in the semantic conventions.
However that is a discussion for another day. This PR #959 looks good as it is.

**Review by christophe-kamphaus-jemmic — APPROVED** [review]

File: src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsSemanticMetrics.java (hunk 9 above)
Comment (cyrille-leclerc) [reply src/main/java/io/jenkins/plugins/opentelemetry/semconv/JenkinsSemanticMetrics.java:10]: Thanks. I'm aligned with you, I also propose to move forward with the `ci.*` namespace and rename once OTel has standardized the namespace and hopefully the metrics we implemented 🤞

**Review by cyrille-leclerc — COMMENTED** [review]
@kuisathaverat can you please review this PR?

**Review by kuisathaverat — APPROVED** [review]

**Pull request merged**